geoutils/georaster/raster.py

```diff
@@ -195,31 +210,97 @@ def __init__(
 
         :return: A Raster object
         """
+        self.driver: str | None = None
+        self.name: str | None = None
+        self.nodata: int | float | list[int] | list[float] | None = None
+        self.filename: str | None = None
+        self.tags: dict[str, Any] = {}
+
+        self._data: np.ndarray | np.ma.masked_array | None = None
+        self._bands = bands
+        self._masked = masked
+        self._disk_hash: int | None = None
+        self._is_modified = True
+        self._disk_shape: tuple[int, int, int] | None = None
+        self._disk_indexes: tuple[int] | None = None
+        self._disk_dtypes: tuple[str] | None = None
+
+        # This is for Raster.from_array to work.
+        if isinstance(filename_or_dataset, dict):
+            self.data = filename_or_dataset["data"]
+            self.transform: rio.transform.Affine = filename_or_dataset["transform"]
+            self.crs: rio.crs.CRS = filename_or_dataset["crs"]
+            self.nodata = filename_or_dataset["nodata"]
+            for key in filename_or_dataset:
+                if key in ["data", "transform", "crs", "nodata"]:
+                    continue
+                setattr(self, key, filename_or_dataset[key])
+            return
+
         # If Raster is passed, simply point back to Raster
         if isinstance(filename_or_dataset, Raster):
             for key in filename_or_dataset.__dict__:
                 setattr(self, key, filename_or_dataset.__dict__[key])
             return
         # Image is a file on disk.
-        elif isinstance(filename_or_dataset, str):
-            # Save the absolute on-disk filename
-            self.filename = os.path.abspath(filename_or_dataset)
-            if as_memfile:
-                # open the file in memory
-                memfile = MemoryFile(open(filename_or_dataset, "rb"))
-                # Read the file as a rasterio dataset
-                self.ds = memfile.open()
+        elif isinstance(filename_or_dataset, (str, rio.io.DatasetReader, rio.io.MemoryFile)):
+
+            with ExitStack():
+                if isinstance(filename_or_dataset, str):
+                    ds: rio.io.DatasetReader = rio.open(filename_or_dataset)
+                    self.filename = filename_or_dataset
+                elif isinstance(filename_or_dataset, rio.io.DatasetReader):
+                    ds = filename_or_dataset
+                    self.filename = filename_or_dataset.files[0]
+                else:
+                    ds = filename_or_dataset.open()
+                    self.filename = None
+
+                self.transform = ds.transform
+                self.crs = ds.crs
+                self.nodata = ds.nodata
+                self.name = ds.name
+                self.driver = ds.driver
+                self.tags.update(ds.tags())
+
+                self._disk_shape = (ds.count, ds.height, ds.width)
```

Comment:
Why rename them and use the underscore prefix? Those are metadata that can be useful to the user.

Reply:
By "them" I mean shape, indexes and dtypes.

Reply:
Nevermind, I see those are defined as properties!